include/dlaf/matrix/print_numpy.h

```diff
@@ -0,0 +1,117 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2020, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <string>
+
+#include "dlaf/common/index2d.h"
+#include "dlaf/common/range2d.h"
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/matrix/tile.h"
+#include "dlaf/types.h"
+
+namespace dlaf {
+namespace matrix {
+
+namespace internal {
+
+// For reference see https://numpy.org/doc/stable/user/basics.types.html
+// Moreover, considering the small number of digits in the output,
+// just single-precision types are used.
+template <class T>
+struct numpy_datatype {
+  static constexpr auto typestring = "single";
+};
+
+template <class T>
+struct numpy_datatype<std::complex<T>> {
+  static constexpr auto typestring = "csingle";
+};
+
+template <class T>
+std::string numpy_value(const T& value) {
+  std::ostringstream os;
+  os << value;
+  return os.str();
+}
+
+template <class T>
+std::string numpy_value(const std::complex<T>& value) {
+  std::ostringstream os;
+  const auto imag = std::imag(value);
+  const auto sign = (imag < 0) ? '-' : '+';
+  os << std::real(value) << sign << std::abs(imag) << "j";
+  return os.str();
+}
+}
+
+/// Print a tile as a numpy array
+template <class Stream, class T>
+Stream& print_numpy(Stream& os, const Tile<const T, Device::CPU>& tile) {
```

Comment:
Wouldn't it be better to name the function `print` and have an argument for the format of the output?

Reply:
Would you go for a template parameter or a tag parameter?

What quickly comes to my mind is:
```cpp
enum class FORMAT{
  NUMPY
};

// usage
print<FORMAT::NUMPY>(std::cout, tile);
```

or
```cpp
struct numpy {}; // optionally in a dedicate namespace `format`

//usage
print(std::cout, tile, numpy{});
```

The first one at first look appears more handy, since you can do something like `using print_numpy = print<numpy>`.

Do you have any suggestion/alternative?

Reply:
A third alternative can be
```
enum class Format{
  numpy
};

// usage
print(std::cout, tile, Format::numpy);
```
Not sure about which is the better structure.

Reply:
That's true! I quickly thought about this and a proposal can be:
- reorder parameters like `print(tile, stream)` so that the stream can have a default parameter (e.g. `std::cout`)
- with the previous point in mind, I would use the format as template parameter. This would allow to specify or use the default value for stream and format, independently. In fact, if both of them are classic function parameters, depending on the order, if we want to specify the second one, we are forced to specify also the first one (assuming we want default values).

So, this proposal would be something like
```cpp
// enum/struct for describing available formats

print(tile);            // default format and default stream
print(tile, std::cout); // default format and output on cout
print<numpy>(tile);     // numpy format and default stream
print(tile, std::cout); // default format and output on cout
```

Reply:
An addition/correction to my previous comment. Since it is already a templated function, adding a default parameter as a template would mean partial specialization for it, which is not allowed. On the other hand, we could use enable_if, but I think it is a big machinery not worth it.

I would go for not having a default format, and have it as a normal parameter before the stream (which could still be optional). And having the format as parameter, I would opt for struct (possibly in a namespace) instead of enum, so we can use overloading for selecting different format implementation.